lang/cs/Org.Apache.REEF.Network/Elastic/Driver/Default/DefaultElasticContext.cs

```diff
@@ -0,0 +1,428 @@
+        /// It gets a failed task as input and in response it produces zero or more failure events.
+        /// </summary>
+        /// <param name="task">The failed task</param>
+        /// <param name="failureEvents">A list of events encoding the type of actions to be triggered so far</param>
+        /// <exception cref="Exception">If the task failure cannot be properly handled</exception>
+        public void OnTaskFailure(IFailedTask value, ref List<IFailureEvent> failureEvents)
+        {
+            var task = value.Id;
+            _nameServer.Unregister(task);
+        }
+
+        /// <summary>
+        /// Used to react when a timeout event is triggered.
+        /// It gets a failed task as input and in response it produces zero or more failure events.
+        /// </summary>
+        /// <param name="alarm">The alarm triggering the timeput</param>
+        /// <param name="msgs">A list of messages encoding how remote Tasks need to reach</param>
+        /// <param name="nextTimeouts">The next timeouts to be scheduled</param>
+        public void OnTimeout(Alarm alarm, ref List<IElasticDriverMessage> msgs, ref List<ITimeout> nextTimeouts)
+        {
+        }
+
+        /// <summary>
+        /// When a new failure state is reached, this method is used to dispatch
+        /// such event to the proper failure mitigation logic.
+        /// It gets a failure event as input and produces zero or more failure response messages
+        /// for tasks (appended into the event).
+        /// </summary>
+        /// <param name="event">The failure event to react upon</param>
+        public void EventDispatcher(ref IFailureEvent @event)
+        {
+            switch ((DefaultFailureStateEvents)@event.FailureEvent)
+            {
+                case DefaultFailureStateEvents.Reconfigure:
+                    var rec = @event as ReconfigureEvent;
+                    OnReconfigure(ref rec);
+                    break;
+
+                case DefaultFailureStateEvents.Reschedule:
+                    var res = @event as RescheduleEvent;
+                    OnReschedule(ref res);
+                    break;
+
+                case DefaultFailureStateEvents.Stop:
+                    var stp = @event as StopEvent;
+                    OnStop(ref stp);
+                    break;
+
+                default:
+                    OnFail();
+                    break;
+            }
+        }
+
+        #endregion Failure Response
+
+        #region Default Failure event Response
+
+        /// <summary>
+        /// Mechanism to execute when a reconfigure event is triggered.
+        /// <paramref name="reconfigureEvent"/>
+        /// </summary>
+        public void OnReconfigure(ref ReconfigureEvent reconfigureEvent)
+        {
+            lock (_statusLock)
+            {
+                _failureStatus = _failureStatus.Merge(
+                    new DefaultFailureState((int)DefaultFailureStates.ContinueAndReconfigure));
```

Comment:
I really don't like that we have three different representations for the failure state - `int`, `enum`, and `class`/`interface`. That leads to awkward casts and `new`s. I am still thinking about the most elegant way to resolve this; one approach would be to remove the `DefaultFailureStates` enum, and go old Java style, like this:
```C#
public sealed class DefaultFailureState : IFailureState
{
    public static readonly DefaultFailureState Continue = new DefaultFailureState(0);
    public static readonly DefaultFailureState ContinueAndReconfigure = new DefaultFailureState(1);
    // ...
```
then here we can just write
```C#
_failureStatus = _failureStatus.Merge(DefaultFailureState.ContinueAndReconfigure);
```
but I am not 100% sure... maybe we can keep the `enum`, ditch the `IFailureState` interface, and use implicit operators to convert between the class and enum..

Reply:
I agree that is weird. I personally like the `IFailureState` interface because it allows to implement different failure states and related ways of merging them. If I have to chose, I would prefer to maintain the interface and implement the default behavior in a different way (e.g., as you were suggesting, without the `enum`).